Packages/com.unity.inputsystem/InputSystem/Editor/InputLayoutCodeGenerator.cs

```diff
@@ -0,0 +1,335 @@
+        /// <param name="defines">Null/empty or a valid expression for an #if conditional compilation statement.</param>
+        /// <param name="namePrefix">Prefix to prepend to the type name of <paramref name="layoutName"/>.</param>
+        /// <param name="visibility">C# access modifier to use with the generated class.</param>
+        /// <param name="namespace">Namespace to put the generated class in. If <c>null</c>, namespace of type behind <paramref name="layoutName"/> will be used.</param>
+        /// <returns>C# source code for a precompiled version of the device layout.</returns>
+        /// <remarks>
+        /// The code generated by this method will be many times faster than the reflection-based <see cref="InputDevice"/>
+        /// creation normally performed by the input system. It will also create less GC heap garbage.
+        ///
+        /// The downside to the generated code is that the makeup of the device is hardcoded and can no longer
+        /// be changed by altering the <see cref="InputControlLayout"/> setup of the system.
+        ///
+        /// Note that it is possible to use this method with layouts generated on-the-fly by layout builders such as
+        /// the one employed for <see cref="HID"/>. However, this must be done at compile/build time and can thus not
+        /// be done for devices dynamically discovered at runtime. When this is acceptable, it is a way to dramatically
+        /// speed up the creation of these devices.
+        /// </remarks>
+        /// <seealso cref="InputSystem.RegisterPrecompiledLayout{T}"/>
+        public static string GenerateCodeForDeviceLayout(string layoutName, string defines = null, string namePrefix = "Fast", string visibility = "public", string @namespace = null)
+        {
+            if (string.IsNullOrEmpty(layoutName))
+                throw new ArgumentNullException(nameof(layoutName));
+
+            // Produce a device from the layout.
+            var device = InputDevice.Build<InputDevice>(layoutName, noPrecompiledLayouts: true);
+
+            // Get info about base type.
+            var baseType = device.GetType();
+            var baseTypeName = baseType.Name;
+            var baseTypeNamespace = baseType.Namespace;
+
+            // Begin generating code.
+            var writer = new InputActionCodeGenerator.Writer
+            {
+                buffer = new StringBuilder()
+            };
+
+            writer.WriteLine(CSharpCodeHelpers.MakeAutoGeneratedCodeHeader("com.unity.inputsystem:InputLayoutCodeGenerator",
+                InputSystem.version.ToString(),
+                $"\"{layoutName}\" layout"));
+
+            // Defines.
+            if (defines != null)
+            {
+                writer.WriteLine($"#if {defines}");
+                writer.WriteLine();
+            }
+
+            if (@namespace == null)
+                @namespace = baseTypeNamespace;
+
+            writer.WriteLine("using UnityEngine.InputSystem;");
```

Comment:
Feels like we could do with a template file for this stuff.

Reply:
Maybe. Bit reluctant as it will blob up what is now one single self-contained, very obvious and easy to change file into a something that needs to be modified in multiple places and is far less straightforward to digest and modify.